How To - Work with local models

haywoodsloan opened the pull request:
## Summary

Add the how-to doc: _Work with local models_


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/ai/how-to/work-with-local-models.md](https://github.com/dotnet/docs/blob/64e360c035f85a1f9bc2f38f42d24676e57f181b/docs/ai/how-to/work-with-local-models.md) | [Use custom and local AI models with the Semantic Kernel SDK](https://review.learn.microsoft.com/en-us/dotnet/ai/how-to/work-with-local-models?branch=pr-en-us-40470) |


<!-- PREVIEW-TABLE-END -->

haywoodsloan committed:
start doc

haywoodsloan committed:
add sample code

haywoodsloan committed:
add invoke prompt example for chat completion

haywoodsloan committed:
finish first draft

haywoodsloan committed:
some rewording

haywoodsloan commented:
@catbutler ready for your review

catbutler commented on docs/ai/how-to/work-with-local-models.md:
```diff
@@ -0,0 +1,59 @@
+
+## Related content
+
+* [What is Semantic Kernel](/semantic-kernel/overview/)
```
Maybe a link about plugins? Also, if there's decision-support content re: "Kernel vs. service class" that would be useful to include.

haywoodsloan replied:
Good suggestion, added a link to the plugins doc! 

I haven't seen any docs on Kernel vs service class, but based on [this section heading](https://learn.microsoft.com/en-us/semantic-kernel/agents/kernel/?tabs=Csharp#the-kernel-is-at-the-center-of-everything) I think using Kernel whenever possible is prefered. Looking at the SDK reference, the Kernel gives the most versibility and functionality (plugins, functions, modle agnostic). The service class would be most useful if you need to interact with your model directly (tasks specific to the model type). 

I think going forward I'm going to refernce Kernel functions as much as possible, but want to show how you can retreive your custom service class from the Kernel in these samples.

catbutler reviewed (changes requested):
Finally got back in the PR swing of things, so you get proper suggestions now! 😃  Also found more metadata guidelines for titles and descriptions. 

I like this doc's elegant structure and supporting snippets. It did leave me wondering how to use the steps to integrate a given model - that might be me failing to see something, but if that **isn't** covered I think it should be. I also left some specific suggestions inline. Cheers!

haywoodsloan commented on docs/ai/how-to/snippets/semantic-kernel/LocalModelExamples.cs:
```diff
@@ -0,0 +1,123 @@
+        // <useTextService>
+        var executionSettings = new PromptExecutionSettings
+        {
+            // Add execution settings, such as the ModelID and ExtensionData
```
I could add some example/placeholder initilizer, I left them out at first for brevity. Don't mind adding them though.

haywoodsloan replied:
Added some sample initilizers demonstrating how you might pass extra settings via the ExtensionData property

haywoodsloan committed:
Use a lazy-init pattern for the Attributes property

catbutler committed:
revised metadata and content to better match guidelines

catbutler committed:
revised metadata and content to better match guidelines

catbutler committed:
Merge branch 'shaywood/howto-local-models' of https://github.com/haywoodsloan/docs into pr/haywoodsloan/40470

haywoodsloan committed:
Revert "revised metadata and content to better match guidelines"

This reverts commit 2925c10074bf08527db25bce7e69353f5c4189bb.

haywoodsloan commented on docs/ai/how-to/work-with-local-models.md:
That's fair I might have abridged too much and led to a "draw the rest of the owl" situation. I'll add some more mock code to the service classes that gives a basic example of interacting with a custom model to send/recieve completions.

haywoodsloan replied:
Updated the snippets to have more mock code!

haywoodsloan committed:
pr feedback

haywoodsloan committed:
update code samples to show more mock code around the model interactions

haywoodsloan commented:
> Finally got back in the PR swing of things, so you get proper suggestions now! 😃 Also found more metadata guidelines for titles and descriptions.
> 
> I like this doc's elegant structure and supporting snippets. It did leave me wondering how to use the steps to integrate a given model - that might be me failing to see something, but if that **isn't** covered I think it should be. I also left some specific suggestions inline. Cheers!

@catbutler thanks for the review! You have a good point that the code samples could show some more mock code about interacting with your model from the service classes. I updated the code snippets with some more mocks. 

Ready for a second pass!

BillWagner commented on docs/ai/how-to/snippets/semantic-kernel/LocalModelExamples.cs:
```diff
@@ -0,0 +1,127 @@
+        Kernel kernel = builder.Build();
+        // </addTextService>
+
+        return builder.Build();
```
Hmm. Is this idempotent? Maybe this instead:

```suggestion
        return kernel;
```

BillWagner commented on docs/ai/how-to/snippets/semantic-kernel/LocalModelExamples.cs:
```diff
@@ -0,0 +1,127 @@
+
+        // Send a prompt to your model directly through the Kernel
+        string prompt = "Please list three services offered by Azure";
+        string? response = await kernel.InvokePromptAsync<string>(prompt);
```
OK, more an API question, but it seems odd to me that this could return `null`. Maybe explain what conditions would produce a null response?

haywoodsloan replied:
Unfortunately, the documentation doesn't specify when a null value is returned: [Ref Docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.semantickernel.kernelextensions.invokepromptasync?view=semantic-kernel-dotnet#microsoft-semantickernel-kernelextensions-invokepromptasync-1(microsoft-semantickernel-kernel-system-string-microsoft-semantickernel-kernelarguments-system-string-microsoft-semantickernel-iprompttemplatefactory)).

@elbruno do you know what might make the return null?

haywoodsloan replied:
My initial guess is that `InvokePromptAsync<TValue>(prompt)` can return null if the output of the model can't be parsed as `TValue`

elbruno replied:
I'm testing locally this to understand more about it. I also think that if the model can't be accessed, it will return null. We can add a note here about it.

haywoodsloan replied:
Added a comment to explain that! Thanks!!

BillWagner commented on docs/ai/how-to/snippets/semantic-kernel/LocalModelExamples.cs:
```diff
@@ -0,0 +1,127 @@
+        // Send a string representation of the chat history to your model directly through the Kernel
+        // This uses a special syntax to denote the role for each message
+        // For more information on this syntax see https://learn.microsoft.com/en-us/semantic-kernel/prompts/your-first-prompt?tabs=Csharp#6-using-message-roles-in-chat-completion-prompts
+        string prompt = @"
+        <message role=""system"">the initial system message for your chat history</message>
+        <message role=""user"">the user's initial message</message>";
```
Yeah, this gets a lot easier with raw string literals:

```suggestion
        string prompt = """
        <message role="system">the initial system message for your chat history</message>
        <message role="user">the user's initial message</message>
        """;
```

BillWagner reviewed (approved):
This LGTM @haywoodsloan 

I had a few comments to consider before you merge, an then it's good.

gewarren commented on docs/ai/how-to/work-with-local-models.md:
```diff
@@ -0,0 +1,60 @@
+---
+title: "Use Custom and Local AI Models with the Semantic Kernel SDK for .NET"
+description: "Learn how to use custom or local models for text generation and chat completions in Semantic Kernel SDK for .NET."
```
```suggestion
titleSuffix: ""
description: "Learn how to use custom or local models for text generation and chat completions in Semantic Kernel SDK for .NET."
```

gewarren commented on docs/ai/how-to/work-with-local-models.md:
```diff
@@ -0,0 +1,60 @@
+
+---
+
+# Use Custom and Local AI Models with the Semantic Kernel SDK
```
```suggestion
# Use custom and local AI models with the Semantic Kernel SDK
```

gewarren reviewed (approved):
When do you plan to add the new docs to the TOC in the working branch?

haywoodsloan commented:
> When do you plan to add the new docs to the TOC in the working branch?

As soon as that PR is merged, we're waiting to complete any other PRs until then. I'll update this PR to include the TOC addition.

haywoodsloan committed:
Apply suggestions from code review

Co-authored-by: Bill Wagner <wiwagn@microsoft.com>
Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

catbutler committed:
Merge branch 'shaywood/howto-local-models' of https://github.com/haywoodsloan/docs into pr/haywoodsloan/40470

haywoodsloan committed:
Revert "revised metadata and content to better match guidelines"

This reverts commit 26c3037d047002d06679518ffa109ce40497f3c1.

haywoodsloan committed:
Merge branch 'new-conceptual-ai' of https://github.com/haywoodsloan/docs into shaywood/howto-local-models

haywoodsloan committed:
Add TOC entry

haywoodsloan commented:
> > When do you plan to add the new docs to the TOC in the working branch?
> 
> As soon as that PR is merged, we're waiting to complete any other PRs until then. I'll update this PR to include the TOC addition.

@gewarren TOC update is now added to this PR!

gewarren commented on docs/ai/toc.yml:
```diff
@@ -4,6 +4,8 @@ items:
 - name: How-to articles
-  items: []
+  items:
+  - name: Use custom and local AI models with the Semantic Kernel SDK
```
```suggestion
  - name: Use AI models with the Semantic Kernel SDK
```

haywoodsloan replied:
I would keep "custom and local" because this doc doesn't cover how to use the "standard" models (OpenAI, AOAI). So the important distinction with this article is that it explains how to integrate something custom that isn't part of SK out of the box

gewarren reviewed (approved):
One suggestion just to shorten the TOC entry if you agree.

haywoodsloan committed:
Merge branch 'new-conceptual-ai' of https://github.com/dotnet/docs into shaywood/howto-local-models

haywoodsloan committed:
add comment to clarify when responses are null

haywoodsloan committed:
fix typo

elbruno commented:
A couple of general notes on the whole doc:

- We need to add a clear description, that accessing custom LLMs (i.e.: Llama2) is also using HTTP calls. That means that the model need to be hosted 1st, using ollama in example. The prerequisites explains this, a link to this ollama blog may help the audience to know how to run local models (https://ollama.com/blog/openai-compatibility)
- In the MyTextGeneration.cs and MyChatCompletion.cs samples, we use MyModel as a mock model (docs/ai/how-to/snippets/semantic-kernel/model/MyModel.cs). It will be nice to mention and explain that this is a mock model. 
- This is a just a suggestion: besides the mock MyModel.cs, I think we will share some value to show an implementation of a real custom model implementing the http requests to a local model. Something like this one (https://github.com/elbruno/semantickernel-localLLMs/blob/main/src/sk-customllm/CustomChatCompletionService.cs). And show the lines on how to change the [var service = new MyChatCompletionService();] to the real custom model.

haywoodsloan committed:
updates based on PR feedback

haywoodsloan commented:
> A couple of general notes on the whole doc:
> 
> * We need to add a clear description, that accessing custom LLMs (i.e.: Llama2) is also using HTTP calls. That means that the model need to be hosted 1st, using ollama in example. The prerequisites explains this, a link to this ollama blog may help the audience to know how to run local models (https://ollama.com/blog/openai-compatibility)
> * In the MyTextGeneration.cs and MyChatCompletion.cs samples, we use MyModel as a mock model (docs/ai/how-to/snippets/semantic-kernel/model/MyModel.cs). It will be nice to mention and explain that this is a mock model.
> * This is a just a suggestion: besides the mock MyModel.cs, I think we will share some value to show an implementation of a real custom model implementing the http requests to a local model. Something like this one (https://github.com/elbruno/semantickernel-localLLMs/blob/main/src/sk-customllm/CustomChatCompletionService.cs). And show the lines on how to change the [var service = new MyChatCompletionService();] to the real custom model.

@elbruno thanks for the feedback! I've incorporated the suggested changes:
- Add details to the introduction that explain most models will be accessed via HTTP and REST APIs, you'll need to host the model in a way that is accessible to your application via HTTP, a link to the Ollama blog as an example of a REST API for completions.
- Removed the mock model and replaced it with HTTP requests
- Added comments to the service initializers to show that's where the user would update the configuration based on their service implementation.

alexwolfmsft commented:
@haywoodsloan One general feedback item for this doc and others - can we try to format code samples so that they render without horizontal scrollbars? We usually break long lines of code onto multiple lines so the user can see it all at once.

haywoodsloan committed:
format code and remove old file

haywoodsloan reviewed (commented):
> @haywoodsloan One general feedback item for this doc and others - can we try to format code samples so that they render without horizontal scrollbars? We usually break long lines of code onto multiple lines so the user can see it all at once. 

Sure thing! I used the CSharpier extension to re-format the code. I can't access the preview links so let me know if it looks okay to you! I did the same thing for my other How-To article as well.

haywoodsloan committed:
move link to new line to prevent scroll
